File: src/spoj/dynamic_programming/mkbudget/MKBUDGET.c
// http://www.spoj.com/problems/MKBUDGET/

#include <stdio.h>
#include <stdint.h>

#define MAX_COUNT 25
#define INFINITY 2000000000L

// array, which represents minimal required amount of workers
// for each month
int64_t workers[MAX_COUNT];

// number of months
int32_t length;

// cost for previous and current months
int64_t cost[2][MAX_COUNT];
int32_t prev;
int32_t curr;

// cost to hire new worker
int64_t hire;
// monthly salary of worker
int64_t salary;
// cost which company pys to worker, when he leaving company
int64_t severance;

int64_t find_minimum_cost() {
	int32_t m; // month index
	int32_t c; // current count index
	int32_t k; // previous count index

	prev = 0;
	curr = 1;

	for(m = 0; m < length; ++m) {
		cost[prev][m] = (workers[0] > workers[m]) 
						? INFINITY 
						: workers[m] * (salary + hire);
	}

	int64_t curr_cost;

	for(m = 1; m < length; ++m) {
		for(c = 0; c < length; ++c) {
			cost[curr][c] = INFINITY;
			if(workers[c] < workers[m]) {
				continue;
			}
			for(k = 0; k < length; ++k) {
				curr_cost = cost[prev][k] + workers[c] * salary;
				if(workers[c] < workers[k]) {
					curr_cost += (workers[k] - workers[c]) * severance;
				}
				if(workers[c] > workers[k]) {
					curr_cost += (workers[c] - workers[k]) * hire;
				}
				cost[curr][c] = (curr_cost > cost[curr][c])
								? cost[curr][c]
								: curr_cost;
			}
		}
		// swap prev and curr
		prev ^= 1;
		curr ^= 1;
	}

	int64_t result = INFINITY;
	for(c = 0; c < length; ++c) {
		result = (result > cost[prev][c])
				 ? cost[prev][c]
				 : result;
	}

	return result;
}

int main() {
	int i;
	int64_t result;
	scanf("%d", &length);

	int case_num = 1;
	
	while(length) {
		scanf("%lld %lld %lld", &hire, &salary, &severance);
		for(i = 0; i < length; ++i) {
			scanf("%lld", &workers[i]);
		}
		result = find_minimum_cost();
		printf("Case %d, cost = $%lld\n", case_num, result);
		scanf("%d", &length);
		case_num++;
	}

	return 0;
}

File: src/spoj/dynamic_programming/mkbudget/mkbudget.py
#!/usr/bin/python

"""
Solution of: http://www.spoj.com/problems/MKBUDGET/
"""
def get_min_budget(month_count, hire, salary, severance):
    INF = 1000000
    length = len(month_count)
    cost = [[0] * length for i in xrange(length)]
    for c in xrange(length):
        cost[0][c] = (month_count[c] * (hire + salary)) if month_count[c] >= month_count[0] else INF
    for m in xrange(1, length):
        for c in xrange(length):
            cost[m][c] = INF
            if month_count[c] < month_count[m]:
                continue
            for k in xrange(length):
                curr = month_count[c] * salary + cost[m - 1][k]
                if month_count[c] < month_count[k]:
                    curr += (month_count[k] - month_count[c]) * severance
                elif month_count[c] > month_count[k]:
                    curr += (month_count[c] - month_count[k]) * hire
                cost[m][c] = min(cost[m][c], curr)
    result = INF
    for c in xrange(length):
        result = min(result, cost[length - 1][c])
    return result

print get_min_budget([10, 9, 11], 400, 500, 600)
print get_min_budget([11, 9, 10, 14, 9, 9, 13, 15], 400, 600, 600)
